ekm/remote_key_manager.go

```diff
@@ -0,0 +1,399 @@
+		getOperatorId:     getOperatorId,
+		operatorPubKey:    operatorPubKey,
+	}, nil
+}
+
+func (km *RemoteKeyManager) AddShare(
+	ctx context.Context,
+	encryptedSharePrivKey []byte,
+	sharePubKey phase0.BLSPubKey,
+) error {
+	shareKeys := ssvsigner.ShareKeys{
+		EncryptedPrivKey: hexutil.Bytes(encryptedSharePrivKey),
+		PublicKey:        sharePubKey,
+	}
+
+	statuses, err := km.remoteSigner.AddValidators(ctx, shareKeys)
+	if err != nil {
+		return fmt.Errorf("add validator: %w", err)
+	}
+
+	// AddValidators validates response length.
+	if statuses[0] != web3signer.StatusImported {
+		return fmt.Errorf("unexpected status %s", statuses[0])
+	}
+
+	if err := km.BumpSlashingProtection(sharePubKey); err != nil {
+		return fmt.Errorf("could not bump slashing protection: %w", err)
+	}
+
+	return nil
+}
+
+func (km *RemoteKeyManager) RemoveShare(ctx context.Context, pubKey phase0.BLSPubKey) error {
+	statuses, err := km.remoteSigner.RemoveValidators(ctx, pubKey)
+	if err != nil {
+		return fmt.Errorf("remove validator: %w", err)
+	}
+
+	// RemoveValidators validates response length.
+	if statuses[0] != web3signer.StatusDeleted {
+		return fmt.Errorf("received status %s", statuses[0])
+	}
+
+	if err := km.RemoveHighestAttestation(pubKey); err != nil {
+		return fmt.Errorf("could not remove highest attestation: %w", err)
+	}
+	if err := km.RemoveHighestProposal(pubKey); err != nil {
+		return fmt.Errorf("could not remove highest proposal: %w", err)
+	}
+
+	return nil
+}
+
+func (km *RemoteKeyManager) SignBeaconObject(
+	ctx context.Context,
+	obj ssz.HashRoot,
+	domain phase0.Domain,
+	sharePubkey phase0.BLSPubKey,
+	slot phase0.Slot,
+	signatureDomain phase0.DomainType,
+) (spectypes.Signature, phase0.Root, error) {
+	epoch := km.netCfg.Beacon.EstimatedEpochAtSlot(slot)
+
+	forkInfo, err := km.getForkInfo(ctx, epoch)
+	if err != nil {
+		return spectypes.Signature{}, phase0.Root{}, fmt.Errorf("get fork info: %w", err)
+	}
+
+	req := web3signer.SignRequest{
+		ForkInfo: forkInfo,
+	}
+
+	switch signatureDomain {
+	case spectypes.DomainAttester:
+		data, ok := obj.(*phase0.AttestationData)
+		if !ok {
+			return nil, phase0.Root{}, errors.New("could not cast obj to AttestationData")
+		}
+
+		if err := km.IsAttestationSlashable(sharePubkey, data); err != nil {
+			return nil, phase0.Root{}, err
+		}
+
+		if err := km.UpdateHighestAttestation(sharePubkey, data); err != nil {
+			return nil, phase0.Root{}, err
+		}
+
+		req.Type = web3signer.TypeAttestation
+		req.Attestation = data
+
+	case spectypes.DomainProposer:
+		switch v := obj.(type) {
+		case *capella.BeaconBlock, *deneb.BeaconBlock, *electra.BeaconBlock:
+			return nil, phase0.Root{}, fmt.Errorf("web3signer supports only blinded blocks since bellatrix") // https://github.com/Consensys/web3signer/blob/85ed009955d4a5bbccba5d5248226093987e7f6f/core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/signing/eth2/BlockRequest.java#L29
```

Comment:
Does this mean we can't make full blocks with remote signer? what if we don't have any MEV setup?

Reply:
adding a link to a previous discussion because it got lost in the history: https://github.com/ssvlabs/ssv/pull/2028#discussion_r1982332723

Reply:
Actually, both blinded and non-blinded blocks need to be hashed before passing to web3signer (it expects `BeaconBlockHeader`) type, so it's the same process for both blinded and non-blinded and I was wrong here. I pushed the fix